### PR TITLE
test: migrate v2 contracts unit tests to foundry

### PR DESCRIPTION
**Pull request by skosito**
- migrates unit tests for v2 contracts from hardhat to foundry
- GatewayEVMUniswap is not migrated, it is tricky and everything that uniswap deploy npm package does we would need to do manually (basically this whole repo: https://github.com/trashpirate/uniswapV2-deploy, which means manually coping all uniswap files to fix uniswap pair issue on local network) - can this remain hardhat or be part of worker to test this flow?

- testFoundry folder separated from test, to better separate hardhat and foundry due to some unexpected errors
- introduces openzeppelin foundry upgrade, and modifies tests to deploy gateway and proxy properly
- fixes build, test.yml was hanging again, it seems that moving to nodev4 and checkoutv4 did the trick
EDIT: it still hangs, this is to be resolved in separate issue as it is ongoing and flaky for some time and not related to this PR

- various fixes for issues discovered through debugging (config changes and so on)

- nice to add in separate PR, fix lint warnings and some kind of formatter for solidity files, eg: https://book.getfoundry.sh/reference/config/formatter

## Summary by CodeRabbit

- **New Features**
  - Introduced new test contracts for GatewayEVM and its interactions with various other contracts.

- **Bug Fixes**
  - Updated workflow files to use the latest versions of GitHub actions.
  - Improved Solidity version compatibility by updating pragma directives.

- **Chores**
  - Added new submodule declaration for openzeppelin-foundry-upgrades.
  - Updated `package.json` scripts for improved linting and testing.
  - Enhanced `.gitignore` to exclude coverage files.

- **Tests**
  - Added comprehensive test cases for contracts in EVM and ZEVM environments.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,6 @@
 [submodule "lib/forge-std"]
 	path = lib/forge-std
 	url = https://github.com/foundry-rs/forge-std
+[submodule "lib/openzeppelin-foundry-upgrades"]
```
Comment (fbac) [inline .gitmodules:4]: nit: why is this submodule needed?

File: foundry.toml [hunk 2]
```diff
@@ -2,8 +2,14 @@
 src = 'contracts'
 out = 'out'
 libs = ['node_modules', 'lib']
-test = 'test'
+test = 'testFoundry'
```
Comment (fbac) [inline foundry.toml:5]: Why moving foundry to testFoundry? Can't have test/fuzz, test/prototypes and test/foundry for example?

Comment (fbac) [reply foundry.toml:5]: Understood. Isn't it possible to move them to test/foundry and exclude the path from hardhat?

**Review by lumtis — COMMENTED** [review]
Looks good to me, I just would like to check the coverage.
I'm doing a branch right now to update the coverage report

**Review by lumtis — COMMENTED** [review]
Would be great if we can merge https://github.com/zeta-chain/protocol-contracts/pull/231 to it

The coverage looks good https://app.codecov.io/github/zeta-chain/protocol-contracts
I think we're mainly missing tests for ERC20Custody withdraw

**Review by lumtis — APPROVED** [review]

**Review by fbac — APPROVED** [review]

**Pull request merged**